vpr/src/route/rr_graph_uxsdcxx.h

```diff
@@ -0,0 +1,3873 @@
+inline void load_edge_required_attributes(const pugi::xml_node& root, unsigned int* sink_node, unsigned int* src_node, unsigned int* switch_id);
+template<class T>
+inline void load_rr_edges(const pugi::xml_node& root, T& out, const void* data, void* iter);
+template<class T>
+inline void load_rr_graph(const pugi::xml_node& root, T& out, const void* data, void* iter);
+
+/* Declarations for internal write functions for the complex types. */
+template<class T>
+inline void write_channels(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_switch(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_switches(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_segment(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_segments(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_pin(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_pin_class(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_block_type(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_block_types(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_grid_locs(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_meta(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_metadata(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_node(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_rr_nodes(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_edge(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_rr_edges(T& in, std::ostream& os, const void* data, void* iter);
+template<class T>
+inline void write_rr_graph(T& in, std::ostream& os, const void* data, void* iter);
+
+/* Load function for the root element. */
+template<class T>
+inline pugi::xml_parse_result load_rr_graph_xml(T& out, std::istream& is) {
+    static_assert(std::is_base_of<RrGraphBase, T>::value, "Base class not derived from RrGraphBase");
+    pugi::xml_document doc;
+    pugi::xml_parse_result result = doc.load(is);
+    if (!result) return result;
+    for (pugi::xml_node node = doc.first_child(); node; node = node.next_sibling()) {
+        if (std::strcmp(node.name(), "rr_graph") == 0) {
+            /* If errno is set up to this point, it messes with strtol errno checking. */
+            errno = 0;
+            load_rr_graph(node, out, nullptr, nullptr);
+        } else
+            throw std::runtime_error("Invalid root-level element " + std::string(node.name()));
```

Comment:
The generated XML reader is actually based on the pugixml parser.  What the generated variant does is handle the string parsing in a more efficient way.  @duck2 can expand if you are wonderommiteding.

One of the future pieces of work would be able to support a SAX parser, so the full DOM can be omited.

Reply:
There are three speed advantages of the generated parser:
1. It goes over DOM subtrees only once in order to check correctness of model groups. A model group in XSD 1.0 is:
a) a tree of "sequence" or "choice"s, describing a regular language of XML tags
b) "all", requiring all given XML tags under it to be present
Instead of checking the tree many times to see if each sub-element under an element exists, it runs a state machine or a state vector over a subtree and then if it ends up in a wrong state, it can emit an error message containing which element is non-existent or in the wrong place.
2. It goes over attributes of each element only once, exactly like described in 1).
3. It uses trie's to tokenize the element, attribute and enumeration names, instead of many strcmp's.

Reply:
Do you have more questions on this front?